compiler/rustc_mir_transform/src/lower_intrinsics.rs

```diff
@@ -191,6 +192,35 @@ impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
                             terminator.kind = TerminatorKind::Goto { target };
                         }
                     }
+                    sym::option_payload_ptr => {
+                        if let (Some(target), Some(arg)) = (*target, args[0].place()) {
+                            let ty::RawPtr(ty::TypeAndMut { ty: dest_ty, .. }) =
+                                destination.ty(local_decls, tcx).ty.kind()
+                            else { return; };
```

Comment:
Is return what we want here or would some kind of `bug!` be better? I'm mostly just going based on what the other branches seem to do where there's not really this early return pattern.

Reply:
I can do `bug!()` instead. Not that it should matter, as the type check should make that code path unreachable unless someone comes up with a way to bypass typeck.